robosherlock/src/flowcontrol/src/RSAggregateAnalysisEngine.cpp

```diff
@@ -191,8 +191,8 @@ void RSAggregateAnalysisEngine::processOnce(std::vector<std::string> &designator
   outInfo("executing analisys engine: " << name_);
   try
   {
-    UnicodeString ustrInputText;
-    ustrInputText.fromUTF8(name_);
+    UnicodeString ustrInputText = name_.c_str();
+    //ustrInputText.fromUTF8(name_); // resulted in an empty ustrInputText
     cas_->setDocumentText(uima::UnicodeStringRef(ustrInputText));
```

Comment:
I don't like the way we set this here. We are abusing this method, and from the RS side it is not clear that this is the name of the AAE working on the CAS. I think we should look at the implementation of the ``cas_->setDocumentText(uima::UnicodeStringRef(ustrInputText));`` method from uimacpp and replicate this with a different name. We have already made a lot of changes in that lib anyway (e.g. made it a ROS package :D )

Reply:
This seems to be a little harder. I tried adding the method but that led into segfaults.
Basically you need three methods: One getter, one setter and a copy method that will memcpy the data from the UnicodeStringRef into the a internal uchar pointer.
When using the set Method, it will segfault in the copyIdentifierString, respectively copyDocumentString method.

Do you know if its save to simply put a std::string into the CAS? With that we could easily add getters and setters..
 
Here is a diff of my stash on trying to replace setDocumentText 
diff --git a/uimacpp_ros/src/cas/cas.cpp b/uimacpp_ros/src/cas/cas.cpp
index 1b33bd8..20a7be1 100644
--- a/uimacpp_ros/src/cas/cas.cpp
+++ b/uimacpp_ros/src/cas/cas.cpp
@@ -706,6 +706,18 @@ namespace uima {
     thisSofa.setLocalSofaData(text);
   }
 
+  void CAS::setIdentifierString(UnicodeStringRef const identifier) {
+    if (isbaseCas) {
+      invalidBaseCasMethod();
+    }
+    SofaFS thisSofa = getSofa();
+    if (!thisSofa.isValid()) {
+      thisSofa = createInitialSofa(UnicodeStringRef("rsidentifier"));
+    }
+    copyIdentifierString(identifier);
+    thisSofa.setLocalSofaData(identifier);
+  }
+
   // internal use only
   void CAS::setDocTextFromDeserializtion(UnicodeStringRef text) {
     copyDocumentString(text);
@@ -733,6 +745,13 @@ namespace uima {
     return UnicodeStringRef(iv_cpDocument, iv_uiDocumentLength);
   }
 
+  UnicodeStringRef CAS::getIdentifierString()const {
+    if (isbaseCas) {
+      return UnicodeStringRef();
+    }
+    return UnicodeStringRef(iv_cpIdentifier, iv_uiIdentifierLength);
+  }
+
   void CAS::createDocumentAnnotation() {
     if (!isbaseCas) {
       assert( EXISTS(iv_heap) );
@@ -1085,6 +1104,23 @@ namespace uima {
     iv_uiDocumentLength = len;
   }
 
+  void CAS::copyIdentifierString(UnicodeStringRef pDoc) {
+    // must have a 2nd copy of the document because the copy in the stringHeap will
+    // move when the stringHeap grows, unlike the older segmented heap.
+    if (pDoc.length() == 0) {
+      return;
+    }
+    if (iv_copyOfIdentifier)
+      delete [] iv_copyOfIdentifier;
+    size_t len = pDoc.length();
+    iv_copyOfIdentifier = new UChar [len + 1];
+    memcpy(iv_copyOfIdentifier, pDoc.getBuffer(), 2*len);
+    // The internal document must be 0 terminated!
+    iv_copyOfIdentifier[len] = 0;
+    iv_cpIdentifier = iv_copyOfIdentifier;
+    iv_uiIdentifierLength = len;
+  }
+
   void CAS::refreshCachedTypes() {
     assert( EXISTS(iv_heap) );
     iv_utDocumentType =
diff --git a/uimacpp_ros/src/cas/include/uima/cas.hpp b/uimacpp_ros/src/cas/include/uima/cas.hpp
index a36c310..30d5547 100644
--- a/uimacpp_ros/src/cas/include/uima/cas.hpp
+++ b/uimacpp_ros/src/cas/include/uima/cas.hpp
@@ -194,6 +194,7 @@ namespace uima {
     void registerView(SofaFS);
     void updateDocumentAnnotation( );
     void copyDocumentString(UnicodeStringRef);
+    void copyIdentifierString(UnicodeStringRef);
     void refreshCachedTypes();
     void createDocumentAnnotation( void );
     void pickupDocumentAnnotation( );
@@ -239,6 +240,9 @@ namespace uima {
     UChar const *               iv_cpDocument;
     size_t                      iv_uiDocumentLength;
     UChar *                     iv_copyOfDocument;
+    UChar const *               iv_cpIdentifier;
+    size_t                      iv_uiIdentifierLength;
+    UChar *                     iv_copyOfIdentifier;
 
     /**
      * Add a copy of <code>crString</code> to the string heap.
@@ -571,6 +575,17 @@ namespace uima {
      *              If the Sofa data has already been set, or View is base CAS.
      */
     void setDocumentText(UnicodeStringRef const text);
+    
+    /**
+     * Set the identifier for this CAS. Once set, Sofa data is immutable, and cannot be set again until the CAS
+     * has been reset.
+     * 
+     * @param identifier
+     *          A string to identify this CAS.
+     * @exception CASException
+     *              If the Sofa data has already been set, or View is base CAS.
+     */
+    void setIdentifierString(UnicodeStringRef const identifier);
 
     /**
      * Set the document text, old style.
@@ -598,6 +613,13 @@ namespace uima {
      */
     virtual UnicodeStringRef getDocumentText() const;
 
+    /**
+     * Get the identifier for this CAS.
+     * 
+     * @return The assigned identifier
+     */
+    virtual UnicodeStringRef getIdentifierString() const;
+
     /**
      * Returns the FeaturesStructure of type document annotation representing
      * the current document.

Reply:
@Sanic Ahh, everythin is always complicated with this lib....Go ahead and try with the std::string. I have no objections. We should eventually ditch this uimacpp library anyway and have our own implementation;

Reply:
Adding a simple std::string as a member seems to work. I'll add push the change in a sec.
However, one has to call setDocumentText() anyway. Without it, the pipeline execution wines about 'can not get default text sofa'.
But atleast we have more clear method for handling the aaename /identying CASes.